# 사용자 함수 정의 및 매핑

---
## 1. 환경 설정

In [16]:
import pandas as pd
import seaborn as sns
import numpy as np

pd.set_option('display.max_columns', 20)
pd.set_option('display.width', 200)

print(f"Pandas 버전: {pd.__version__}")
print(f"Seaborn 버전: {sns.__version__}")

Pandas 버전: 2.2.2
Seaborn 버전: 0.13.2


---
## 3. 함수 매핑 적용 대상 분류

| 대상 | 방법 | 설명 |
|------|------|------|
| **대상 1: Series 개별 원소** | `apply()` 함수 사용 | 1차원 데이터 벡터의 개별 원소 순차 접근 |
| **대상 2: DataFrame 전체 원소** | `applymap()` 함수 사용 | 2차원 전체 원소에 동시 적용 |

기본 문법:
- Series: `Series객체.apply(매핑 함수)`
- DataFrame: `DataFrame객체.applymap(매핑 함수)`

---
## 4. 실습 데이터셋 준비

- 데이터 조작용 패키지 로드: `seaborn`
- `titanic` 내장 데이터셋 변수 할당

In [17]:
# Titanic 데이터셋 로드
import seaborn as sns
titanic = sns.load_dataset('titanic')

print(f"전체 데이터 크기: {titanic.shape}")
print(f"전체 컬럼: {list(titanic.columns)}")
titanic.head()

전체 데이터 크기: (891, 15)
전체 컬럼: ['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


### 4.1 분석 대상 속성 추출

- 대상 컬럼: 생존 여부(survived), 성별(sex), 나이(age), 요금(fare)
- 추출 및 새로운 DataFrame 구성

In [18]:
# 분석 대상 속성 추출
df = titanic.loc[:, ['survived', 'sex', 'age', 'fare']]

print(f"추출 데이터 크기: {df.shape}")
print(f"\n데이터 타입:\n{df.dtypes}")
print(f"\n결측값:\n{df.isnull().sum()}")
print(f"\n상위 5행:")
df.head()

추출 데이터 크기: (891, 4)

데이터 타입:
survived      int64
sex          object
age         float64
fare        float64
dtype: object

결측값:
survived      0
sex           0
age         177
fare          0
dtype: int64

상위 5행:


,survived,sex,age,fare
0,0,male,22.0,7.2500
1,1,female,38.0,71.2833
2,1,female,26.0,7.9250
3,1,female,35.0,53.1000
4,0,male,35.0,8.0500


---
## 5. Series 원소 대상 함수 매핑 분석

### 5.1 `apply()` 함수 적용 원리

- 1차원 데이터 벡터의 **개별 원소 순차 접근**
- 기본 문법: `Series객체.apply(매핑 함수)`

### 5.2 매핑 함수 인자 전달 메커니즘

- Series의 모든 원소를 매핑 함수에 **개별 입력**
- 원소 개수와 동일한 횟수의 **반환값 수신**
- 입력과 동일한 크기(Shape)의 **Series 객체로 최종 반환**

---
## 6. 사용자 정의 함수(Custom Function) 생성

### minus 함수 설계

- **함수명:** `minus(data1, data2)`
- **반환값:** `return data1 - data2`
- **동작:** 두 입력값 간의 차이(Minus) 연산 수행
- `data1`: Series 개별 원소가 자동 할당
- `data2`: 추가 전달 인자

In [19]:
# 사용자 정의 함수: minus
def minus(data1, data2):
    """두 입력값 간의 차이를 반환"""
    return data1 - data2

# 기본 호출 테스트
print(f"minus(50, 10) = {minus(50, 10)}")
print(f"minus(100, 30) = {minus(100, 30)}")

minus(50, 10) = 40
minus(100, 30) = 70


---
## 7. 사용자 함수 매핑 실행: `apply()`


In [20]:
# apply()를 사용한 사용자 함수 매핑 실행
result = df['fare'].apply(minus, data2=10)

print("=" * 60)
print("df['fare'].apply(minus, data2=10) 실행 결과")
print("=" * 60)
print(result)

df['fare'].apply(minus, data2=10) 실행 결과
0      -2.7500
1      61.2833
2      -2.0750
3      43.1000
4      -1.9500
        ...   
886     3.0000
887    20.0000
888    13.4500
889    20.0000
890    -2.2500
Name: fare, Length: 891, dtype: float64


### 7.1 `apply()` 연산 결과 분석

- 모든 fare 원소에 **일괄 -10 연산 완료**
- 인덱스 0 출력: -2.7500 (원래 7.2500 - 10)
- 인덱스 1 출력: 61.2833 (원래 71.2833 - 10)
- 결과 데이터 타입: float64 및 Series
- 크기(Length: 891) 유지

In [21]:
# 결과 분석
print("=" * 60)
print("apply() 연산 결과 분석")
print("=" * 60)
print(f"결과 타입: {type(result)}")
print(f"결과 크기: {len(result)}")
print(f"결과 dtype: {result.dtype}")

print(f"\n 원본 fare 상위 5개:")
print(df['fare'].head())
print(f"\n minus(fare, 10) 결과 상위 5개:")
print(result.head())


apply() 연산 결과 분석
결과 타입: <class 'pandas.core.series.Series'>
결과 크기: 891
결과 dtype: float64

 원본 fare 상위 5개:
0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
Name: fare, dtype: float64

 minus(fare, 10) 결과 상위 5개:
0    -2.7500
1    61.2833
2    -2.0750
3    43.1000
4    -1.9500
Name: fare, dtype: float64


In [22]:
# 다양한 data2 값으로 apply 실행
print("=" * 60)
print("다양한 data2 값으로 apply 실행")
print("=" * 60)

print("\n[ data2=0 (원본 유지) ]")
print(df['fare'].apply(minus, data2=0).head())

print("\n[ data2=50 ]")
print(df['fare'].apply(minus, data2=50).head())

print("\n[ data2=-5 (5를 더하는 효과) ]")
print(df['fare'].apply(minus, data2=-5).head())

다양한 data2 값으로 apply 실행

[ data2=0 (원본 유지) ]
0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
Name: fare, dtype: float64

[ data2=50 ]
0   -42.7500
1    21.2833
2   -42.0750
3     3.1000
4   -41.9500
Name: fare, dtype: float64

[ data2=-5 (5를 더하는 효과) ]
0    12.2500
1    76.2833
2    12.9250
3    58.1000
4    13.0500
Name: fare, dtype: float64


---
## 8. Lambda 함수 매핑

### Lambda 함수의 특징
- 별도 `def` 기반 사용자 함수 정의 과정 **생략**
- 코드 **간결화** 및 메모리 효율성 증대
- 일회성 속성 **실시간 연산에 최적화**

### 기본 문법
```python
df['fare'].apply(lambda x: minus(x, 10))
```
- 변수 `x`: Series의 개별 원소가 자동 할당되는 공간
- 익명 함수 구조를 통한 인자(x, 10) 직접 전달

### 8.1 Lambda 기반 매핑 코드 분석

| 기존 방식 (Standard) | Lambda 방식 (Anonymous) |
|---------------------|------------------------|
| `apply(minus, data2=10)` | `apply(lambda x: minus(x, 10))` |
| def 함수 정의 필요 | 인라인 정의 가능 |

In [23]:
# 기존 방식 vs Lambda 방식 비교
print("=" * 60)
print("기존 방식 vs Lambda 방식 비교")
print("=" * 60)

# 방식 1: 기존 apply (함수명 + 추가 인자)
result_standard = df['fare'].apply(minus, data2=10)

# 방식 2: Lambda 방식
result_lambda = df['fare'].apply(lambda x: minus(x, 10))

print("\n[기존 방식] apply(minus, data2=10):")
print(result_standard.head())

print("\n[Lambda 방식] apply(lambda x: minus(x, 10)):")
print(result_lambda.head())

# 결과 동일성 검증
print(f"\n두 결과가 동일한가? {result_standard.equals(result_lambda)}")

기존 방식 vs Lambda 방식 비교

[기존 방식] apply(minus, data2=10):
0    -2.7500
1    61.2833
2    -2.0750
3    43.1000
4    -1.9500
Name: fare, dtype: float64

[Lambda 방식] apply(lambda x: minus(x, 10)):
0    -2.7500
1    61.2833
2    -2.0750
3    43.1000
4    -1.9500
Name: fare, dtype: float64

두 결과가 동일한가? True


In [24]:
# Lambda만으로 함수 정의 없이 직접 연산
print("=" * 60)
print("Lambda만으로 직접 연산 (def 함수 정의 불필요)")
print("=" * 60)

# def 없이 Lambda만으로 동일한 연산
result_pure_lambda = df['fare'].apply(lambda x: x - 10)
print(result_pure_lambda.head())

# 결과 동일성 검증
print(f"\nminus 함수 결과와 동일한가? {result_standard.equals(result_pure_lambda)}")

Lambda만으로 직접 연산 (def 함수 정의 불필요)
0    -2.7500
1    61.2833
2    -2.0750
3    43.1000
4    -1.9500
Name: fare, dtype: float64

minus 함수 결과와 동일한가? True


---
## 9. DataFrame 전체 원소 대상 함수 매핑 분석

### 9.1 `applymap()` 함수 개요

- `applymap()` 함수 적용 원리
- 2차원 형태 데이터의 **전체 원소에 동시 접근**
- 기본 문법: `DataFrame객체.applymap(매핑 함수)`

### 9.2 `applymap()` 인자 전달 메커니즘

- DataFrame을 구성하는 **모든 개별 원소를 함수에 순차 입력**
- 입력된 2D 형태(Shape)를 **완벽히 유지**
- 입력된 구조의 **새로운 DataFrame 객체로 반환**

### 9.3 DataFrame 2D 연산을 위한 데이터 전처리

- 모든 열을 동시 연산을 위해 **수치형(Numeric) 데이터만 추출**
- 대상: 나이(age), 요금(fare)
- 단일 인자만 전달: `def minus(data): return data - 10`

In [25]:
# DataFrame 2D 연산을 위한 데이터 전처리
import seaborn as sns

titanic = sns.load_dataset('titanic')

# 수치형 데이터만 추출
df = titanic.loc[:, ['age', 'fare']]

print("=" * 60)
print("수치형 데이터 추출 (applymap용)")
print("=" * 60)
print(f"Shape: {df.shape}")
print(f"dtypes:\n{df.dtypes}")
print(f"\n상위 5행:")
print(df.head())

수치형 데이터 추출 (applymap용)
Shape: (891, 2)
dtypes:
age     float64
fare    float64
dtype: object

상위 5행:
    age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500


In [26]:
# applymap용 단일 인자 함수 정의
# (applymap은 추가 인자 전달 불가 → 단일 인자 함수 사용)
def minus10(data):
    """입력값에서 10을 빼서 반환 (단일 인자)"""
    return data - 10

print(f"minus10(50) = {minus10(50)}")
print(f"minus10(7.25) = {minus10(7.25)}")

minus10(50) = 40
minus10(7.25) = -2.75


### 9.4 전체 원소 매핑 실행: `applymap()`

- 실행 코드: `df.applymap(minus10)`
- age (나이) 컬럼 전체 원소: -10 연산 적용
- fare (요금) 컬럼 전체 원소: -10 연산 동시 적용
- 다중 컬럼 **일괄 동시 적용**

In [27]:
# applymap() 실행
# 참고: Pandas 2.1+ 에서는 applymap()이 map()으로 변경됨
# 호환성을 위해 두 가지 방법 모두 시도

print("=" * 60)
print("df.applymap(minus10) 실행 결과")
print("=" * 60)

try:
    # Pandas 2.1 이상
    result_map = df.map(minus10)
    print("(Pandas 2.1+: df.map() 사용)")
except AttributeError:
    # Pandas 2.0 이하
    result_map = df.applymap(minus10)
    print("(Pandas 2.0 이하: df.applymap() 사용)")

print(result_map)

df.applymap(minus10) 실행 결과
(Pandas 2.1+: df.map() 사용)
      age     fare
0    12.0  -2.7500
1    28.0  61.2833
2    16.0  -2.0750
3    25.0  43.1000
4    25.0  -1.9500
..    ...      ...
886  17.0   3.0000
887   9.0  20.0000
888   NaN  13.4500
889  16.0  20.0000
890  22.0  -2.2500

[891 rows x 2 columns]


### 9.5 `applymap()` 연산 결과 및 특성 분석

- age, fare **두 수치형 컬럼 모두 -10** 변환 확인
- 인덱스 0 결과: age(12.0), fare(-2.7500)
- 결측치 보존: 인덱스 888의 NaN(Not a Number)은 연산 후에도 **NaN으로 안전하게 유지**됨

In [28]:
# 결과 분석
print("=" * 60)
print("applymap() 연산 결과 분석")
print("=" * 60)

print(f"\n결과 Shape: {result_map.shape} (원본과 동일)")
print(f"결과 dtypes:\n{result_map.dtypes}")

# 원본 vs 결과 비교 (상위 5행)
print("\n[원본]")
print(df.head())
print("\n[minus10 적용 결과]")
print(result_map.head())

# 결측치 보존 확인
print(f"\n결측치 보존 확인:")
print(f"  원본 age 결측: {df['age'].isnull().sum()}개")
print(f"  결과 age 결측: {result_map['age'].isnull().sum()}개")
print(f"  → NaN은 연산 후에도 NaN으로 안전하게 유지됨")

applymap() 연산 결과 분석

결과 Shape: (891, 2) (원본과 동일)
결과 dtypes:
age     float64
fare    float64
dtype: object

[원본]
    age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

[minus10 적용 결과]
    age     fare
0  12.0  -2.7500
1  28.0  61.2833
2  16.0  -2.0750
3  25.0  43.1000
4  25.0  -1.9500

결측치 보존 확인:
  원본 age 결측: 177개
  결과 age 결측: 177개
  → NaN은 연산 후에도 NaN으로 안전하게 유지됨


In [29]:
# 결측치가 있는 행 확인
print("=" * 60)
print("결측치가 있는 행 비교 (원본 vs 결과)")
print("=" * 60)
nan_rows = df[df['age'].isnull()].head()
print("\n[원본 - NaN 행]:")
print(nan_rows)
print("\n[결과 - NaN 행]:")
print(result_map.loc[nan_rows.index])
print("\n→ age는 NaN 유지, fare는 정상적으로 -10 적용됨")

결측치가 있는 행 비교 (원본 vs 결과)

[원본 - NaN 행]:
    age     fare
5   NaN   8.4583
17  NaN  13.0000
19  NaN   7.2250
26  NaN   7.2250
28  NaN   7.8792

[결과 - NaN 행]:
    age    fare
5   NaN -1.5417
17  NaN  3.0000
19  NaN -2.7750
26  NaN -2.7750
28  NaN -2.1208

→ age는 NaN 유지, fare는 정상적으로 -10 적용됨


### 9.6 Lambda 연산 및 핵심 요약: `applymap()` + Lambda 매핑 비교

```python
df.applymap(lambda x: minus(x, 10))
```

- `applymap()`에도 Lambda 함수 적용 가능
- 단, `applymap()`에서는 추가 인자 전달이 직접 불가하므로 **Lambda를 통해 우회 전달**

In [30]:
# applymap + Lambda 조합
print("=" * 60)
print("applymap + Lambda 조합")
print("=" * 60)

# 원본 minus 함수(2인자)를 Lambda로 감싸서 applymap에 적용
try:
    result_lambda_map = df.map(lambda x: minus(x, 10))
except AttributeError:
    result_lambda_map = df.applymap(lambda x: minus(x, 10))

print(result_lambda_map.head())

# 기존 결과와 동일한지 검증 (NaN 제외 비교)
print(f"\nminus10 함수 결과와 동일한가? {result_map.equals(result_lambda_map)}")

applymap + Lambda 조합
    age     fare
0  12.0  -2.7500
1  28.0  61.2833
2  16.0  -2.0750
3  25.0  43.1000
4  25.0  -1.9500

minus10 함수 결과와 동일한가? True


In [31]:
# Lambda만으로 다양한 applymap 연산
print("=" * 60)
print("Lambda만으로 다양한 applymap 연산")
print("=" * 60)

# 제곱
try:
    print("\n[1] 제곱 연산:")
    print(df.map(lambda x: x ** 2).head())

    print("\n[2] 소수점 1자리 반올림:")
    print(df.map(lambda x: round(x, 1) if pd.notna(x) else x).head())

    print("\n[3] 2배 변환:")
    print(df.map(lambda x: x * 2).head())
except AttributeError:
    print("\n[1] 제곱 연산:")
    print(df.applymap(lambda x: x ** 2).head())

    print("\n[2] 소수점 1자리 반올림:")
    print(df.applymap(lambda x: round(x, 1) if pd.notna(x) else x).head())

    print("\n[3] 2배 변환:")
    print(df.applymap(lambda x: x * 2).head())

Lambda만으로 다양한 applymap 연산

[1] 제곱 연산:
      age         fare
0   484.0    52.562500
1  1444.0  5081.308859
2   676.0    62.805625
3  1225.0  2819.610000
4  1225.0    64.802500

[2] 소수점 1자리 반올림:
    age  fare
0  22.0   7.2
1  38.0  71.3
2  26.0   7.9
3  35.0  53.1
4  35.0   8.1

[3] 2배 변환:
    age      fare
0  44.0   14.5000
1  76.0  142.5666
2  52.0   15.8500
3  70.0  106.2000
4  70.0   16.1000


---
## 10. `apply()` vs `applymap()`: Pandas 함수 매핑 비교

| 구분 | `apply()` | `applymap()` |
|------|-----------|-------------|
| **타겟 객체** | Series 중심 | DataFrame 전용 |
| **데이터 차원** | 1차원 호출 제어 | 2차원 행렬 동시 제어 |
| **사용 목적** | 단일 속성 정밀 가공 | 다중 속성 일괄 스케일링/변환 |
| **추가 인자 전달** | `apply(func, arg=val)` 직접 전달 가능 | Lambda 우회 전달 필요 |

> **참고:** Pandas 2.1 이상에서는 `applymap()`이 `map()`으로 대체되었습니다.

In [32]:
# apply() vs applymap() 종합 비교 실습
print("=" * 70)
print("apply() vs applymap() 종합 비교")
print("=" * 70)

df_demo = titanic.loc[:, ['age', 'fare']]

# apply(): Series 단위 - 단일 열에 적용
print("\n[apply()] Series 단위: fare 열에만 minus(data2=10) 적용")
result_apply = df_demo['fare'].apply(minus, data2=10)
print(result_apply.head())

# applymap(): DataFrame 전체 - 모든 열에 동시 적용
print("\n[applymap()] DataFrame 전체: age, fare 모두에 minus10 적용")
try:
    result_applymap = df_demo.map(minus10)
except AttributeError:
    result_applymap = df_demo.applymap(minus10)
print(result_applymap.head())

apply() vs applymap() 종합 비교

[apply()] Series 단위: fare 열에만 minus(data2=10) 적용
0    -2.7500
1    61.2833
2    -2.0750
3    43.1000
4    -1.9500
Name: fare, dtype: float64

[applymap()] DataFrame 전체: age, fare 모두에 minus10 적용
    age     fare
0  12.0  -2.7500
1  28.0  61.2833
2  16.0  -2.0750
3  25.0  43.1000
4  25.0  -1.9500


---
## 11. 종합 실습: Titanic 데이터 함수 매핑 파이프라인

In [33]:
# 종합 실습 데이터 준비
df_full = titanic.loc[:, ['survived', 'sex', 'age', 'fare']].copy()
print("종합 실습 데이터:")
print(df_full.head(10))

종합 실습 데이터:
   survived     sex   age     fare
0         0    male  22.0   7.2500
1         1  female  38.0  71.2833
2         1  female  26.0   7.9250
3         1  female  35.0  53.1000
4         0    male  35.0   8.0500
5         0    male   NaN   8.4583
6         0    male  54.0  51.8625
7         0    male   2.0  21.0750
8         1  female  27.0  11.1333
9         1  female  14.0  30.0708


In [34]:
# 실습 1: apply()로 나이 구간 분류
def age_group(age):
    """나이를 구간별로 분류"""
    if pd.isna(age):
        return '미상'
    elif age < 18:
        return '미성년'
    elif age < 40:
        return '청장년'
    elif age < 60:
        return '중년'
    else:
        return '노년'

df_full['나이구간'] = df_full['age'].apply(age_group)
print("나이 구간 분류 결과:")
print(df_full['나이구간'].value_counts())
print()
print(df_full.head(10))

나이 구간 분류 결과:
나이구간
청장년    438
미상     177
중년     137
미성년    113
노년      26
Name: count, dtype: int64

   survived     sex   age     fare 나이구간
0         0    male  22.0   7.2500  청장년
1         1  female  38.0  71.2833  청장년
2         1  female  26.0   7.9250  청장년
3         1  female  35.0  53.1000  청장년
4         0    male  35.0   8.0500  청장년
5         0    male   NaN   8.4583   미상
6         0    male  54.0  51.8625   중년
7         0    male   2.0  21.0750  미성년
8         1  female  27.0  11.1333  청장년
9         1  female  14.0  30.0708  미성년


In [35]:
# 실습 2: apply()로 요금 등급 분류 (Lambda 활용)
df_full['요금등급'] = df_full['fare'].apply(
    lambda x: '1등급' if x >= 50 else ('2등급' if x >= 20 else '3등급')
)
print("요금 등급 분류 결과:")
print(df_full['요금등급'].value_counts())
print()
print(df_full.head(10))

요금 등급 분류 결과:
요금등급
3등급    515
2등급    215
1등급    161
Name: count, dtype: int64

   survived     sex   age     fare 나이구간 요금등급
0         0    male  22.0   7.2500  청장년  3등급
1         1  female  38.0  71.2833  청장년  1등급
2         1  female  26.0   7.9250  청장년  3등급
3         1  female  35.0  53.1000  청장년  1등급
4         0    male  35.0   8.0500  청장년  3등급
5         0    male   NaN   8.4583   미상  3등급
6         0    male  54.0  51.8625   중년  1등급
7         0    male   2.0  21.0750  미성년  2등급
8         1  female  27.0  11.1333  청장년  3등급
9         1  female  14.0  30.0708  미성년  2등급


In [36]:
# 실습 3: apply()로 성별 한글 변환
df_full['성별'] = df_full['sex'].apply(lambda x: '남성' if x == 'male' else '여성')
print("성별 한글 변환 결과:")
print(df_full[['sex', '성별']].head())

성별 한글 변환 결과:
      sex  성별
0    male  남성
1  female  여성
2  female  여성
3  female  여성
4    male  남성


In [37]:
# 실습 4: applymap()으로 수치형 데이터 정규화(0~1 스케일링)
df_numeric = df_full[['age', 'fare']].copy()

# 각 열의 min, max 확인
print("정규화 전:")
print(df_numeric.describe().loc[['min', 'max']])

# Min-Max 정규화 함수
def normalize(col):
    """Series를 0~1 범위로 정규화"""
    return (col - col.min()) / (col.max() - col.min())

# apply()로 열 단위 정규화 (axis=0)
df_normalized = df_numeric.apply(normalize)
print("\n정규화 후:")
print(df_normalized.describe().loc[['min', 'max']])
print()
print(df_normalized.head())

정규화 전:
       age      fare
min   0.42    0.0000
max  80.00  512.3292

정규화 후:
     age  fare
min  0.0   0.0
max  1.0   1.0

        age      fare
0  0.271174  0.014151
1  0.472229  0.139136
2  0.321438  0.015469
3  0.434531  0.103644
4  0.434531  0.015713


In [38]:
# 실습 5: 나이구간별 생존율 분석 (파이프라인 종합)
print("=" * 60)
print("나이구간별 생존율 분석 (함수 매핑 파이프라인 종합)")
print("=" * 60)

survival_by_age = df_full.groupby('나이구간').agg(
    총_승객=('survived', 'count'),
    생존자=('survived', 'sum'),
    생존율=('survived', 'mean'),
    평균_요금=('fare', 'mean')
)
survival_by_age['생존율'] = (survival_by_age['생존율'] * 100).round(1).astype(str) + '%'
survival_by_age['평균_요금'] = survival_by_age['평균_요금'].round(2)
print(survival_by_age)

나이구간별 생존율 분석 (함수 매핑 파이프라인 종합)
      총_승객  생존자    생존율  평균_요금
나이구간                         
노년      26    7  26.9%  43.47
미상     177   52  29.4%  22.16
미성년    113   61  54.0%  31.22
중년     137   54  39.4%  41.48
청장년    438  168  38.4%  32.95
